inappmessaging/src/main/java/com/rakuten/tech/mobile/inappmessaging/runtime/manager/DisplayManager.kt

```diff
@@ -33,23 +43,150 @@ internal interface DisplayManager {
-            if (inAppMessageBaseView != null) {
-                // Removing just the InApp Message from the view hierarchy.
-                val parent = inAppMessageBaseView.parent as ViewGroup
+            return if (id != null) {
+                // id is not null if tooltip
+                removeWithId(activity, id, delay)
+                null
+            } else {
+                if (removeAll) {
+                    removeAllTooltip(activity, delay)
+                }
+                // remove normal campaign
+                activity.findViewById<ViewGroup>(R.id.in_app_message_base_view)?.let { viewGroup ->
+                    scheduleRemoval(delay, viewGroup, activity = activity)
+                    viewGroup.tag
+                }
+            }
+        }
+
+        private fun removeWithId(activity: Activity, id: String?, delay: Int) {
+            activity.findViewById<ViewGroup>(R.id.in_app_message_tooltip_view)?.let {
+                if (it.tag == id) {
+                    scheduleRemoval(delay = delay, view = it, id = id, activity = activity)
+                } else {
+                    scheduleTargetChild(it = it, id = id, delay = delay, activity = activity)
+                }
+            }
+        }
+
+        private fun scheduleTargetChild(it: ViewGroup, id: String?, delay: Int, activity: Activity) {
+            it.parent?.let { parent ->
+                for (i in 0 until (parent as ViewGroup).childCount) {
+                    val child = parent.getChildAt(i)
+                    if (child?.id == R.id.in_app_message_tooltip_view && child.tag == id) {
+                        scheduleRemoval(delay = delay, view = child as ViewGroup, id = id, activity = activity)
+                        break
+                    }
+                }
+            }
+        }
+
+        private fun removeAllTooltip(activity: Activity, delay: Int) {
+            activity.findViewById<ViewGroup>(R.id.in_app_message_tooltip_view)?.parent?.let { viewParent ->
+                val viewList = mutableListOf<View>()
+                for (i in 0 until (viewParent as ViewGroup).childCount) {
+                    val child = viewParent.getChildAt(i)
+                    if (child?.id == R.id.in_app_message_tooltip_view) {
+                        viewList.add(child)
+                    }
+                }
+                for (view in viewList) {
+                    scheduleRemoval(
+                        delay = delay, view = view as ViewGroup, id = view.tag as String?,
+                        activity = activity
+                    )
+                }
+            }
+        }
+
+        private fun scheduleRemoval(delay: Int, view: ViewGroup, id: String? = null, activity: Activity) {
+            if (delay > 0) {
+                Handler(Looper.getMainLooper()).postDelayed(
+                    { removeCampaign(view, id, activity) }, delay * MS_MULTIPLIER
```

Comment:
How did you handle repo and impression update when auto-disappear?

Reply:
Sorry it was't handled yet. Added handling